Adds search functionality to extensions page

belcherj opened the pull request:
This ready to go!

To test: Open the extensions page.

- click around and make sure everything
- try the search

belcherj committed:
[WIP] Adds search functionality to extensions page

belcherj committed:
Clean up search bar spacing

belcherj committed:
Fix css on search bar

belcherj committed:
Change placeholder text in searchbar

belcherj committed:
Add build asset

kloon commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -39,6 +39,65 @@ public static function get_featured() {
+	 *
+	 * @return string transient name
+	 */
+	public static function build_transient_name( $category, $term, $country ) {
```
This function is a bit confusing, it takes 3 mandatory params, but it overrides the $transient_name var every time. So if you specify all 3 params then it will always just return the country transient.

belcherj replied:
Maybe I don't understand the comment:  https://repl.it/repls/JollyWorrisomeBarracuda

It builds a string with all three if given all three.  I can add `= null` as defaults if that is better.

kloon commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -39,6 +39,65 @@ public static function get_featured() {
+		if ( false === $addons ) {
+			$raw_extensions = wp_remote_get(
+				'https://woocommerce.com/wp-json/wccom-extensions/1.0/search' . $parameters,
+				array( 'sslverify' => false )
```
Any reason sslverify false is used here? Since I believe wp 3.7 core includes certs with every release which gets queries when sslverify is true, instead of relying on the php install certs.

belcherj replied:
Ahh, yes that was there from me testing on my local.  will remove!

belcherj replied:
Done

kloon reviewed (dismissed):
Couple of small initial code review changes

belcherj committed:
Remove testing code

belcherj committed:
Fix comments to follow standard

claudiulodro commented:
The UI for this looks good.

I wasn't able to get it to actually do a search. If I search for something it just drops me off on the page I was just on with no visible UI change or anything. Is that because it's not fully set up?

There's a PHP notice in the URL bar when searching in the main extensions section:
`/wp-admin/admin.php?search=ok&page=wc-addons&section=<br+%2F><b>Notice<%2Fb>%3A++Undefined+index%3A+section+in+<b>%2Fsrv%2Fwww%2Fwordpress-default%2Fpublic_html%2Fwp-content%2Fplugins%2Fwoocommerce%2Fincludes%2Fadmin%2Fviews%2Fhtml-admin-page-addons.php<%2Fb>+on+line+<b>38<%2Fb><br+%2F>`

belcherj committed:
Add compiled assets

belcherj committed:
Small css fix with margins

belcherj committed:
Account for user searching on featured tab

belcherj commented:
@claudiulodro I fixed the issue in:  https://github.com/woocommerce/woocommerce/pull/17926/commits/9983352d60ccec299de20866c4f6c22a6df783ca

I Think that adding an "all" category is a better solution and I will add that to the api on the other end.

claudiulodro commented on includes/admin/views/html-admin-page-addons.php:
```diff
@@ -29,15 +29,24 @@
+			<input type="text" name="search" value="" placeholder="Enter a search term and press enter">
+			<input type="hidden" name="page" value="wc-addons">
+			<?php $page_section = isset( $_GET['section'] ) ? $_GET['section'] : ''; ?>
+			<?php $page_section = 'featured' === $page_section ? '' : $page_section; ?>
```
I think this can be simplified:
```
<?php $page_section = ( isset( $_GET['section'] ) && 'featured' !== $page_section ) ? $_GET['section'] : ''; ?>
```

belcherj replied:
Done

claudiulodro commented:
Search works great now! Just a couple notes then I think this looks good to go:
- I think the current search query should be displayed in the input box on search results. Currently there's no visual indication what query the search results are for when performing a search.
- (minor. Can be fixed later since it's not really in the scope of this PR) Float issue (search "test" in the featured panel to reproduce):
<img width="1266" alt="screen shot 2017-12-01 at 9 24 57 am" src="https://user-images.githubusercontent.com/7317227/33495385-13dfc3f4-d67b-11e7-8eb3-aa60c0f56b32.png">

- I don't see any extensions in the "Popular" tab. This may be because it's not quite ready yet or something. Just an observation. Probably not related to this PR.

garymurray reviewed (commented):
A few small things:
- When viewing the search bar on the featured tab - the two elements are touching and out of aligment:

![screen shot 2017-12-04 at 09 51 10](https://user-images.githubusercontent.com/10805911/33541736-e156e556-d8d8-11e7-8c3f-99d465edcef3.png)

- We should look to have some kind of indication on the search results page of what a customer searched for - either as per @claudiulodro comment or we can include it as per this option:

![screen shot 2017-12-04 at 09 56 27](https://user-images.githubusercontent.com/10805911/33541889-8dd8ad3c-d8d9-11e7-977f-319e0c212142.png)

belcherj committed:
Update for API changes

belcherj committed:
Merge Commit

belcherj committed:
Build CSS

belcherj committed:
Fix css and add search query header

belcherj commented:
I have added @garymurray 's suggestion and fixed the css issue he observed.

I'll fix the issue with the float in another PR.  I need to limit length of text and add ellipsis...

claudiulodro commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -40,36 +40,97 @@ public static function get_featured() {
+	 */
+	public static function get_sections() {
+		$addon_sections = get_transient( 'wc_addons_sections' );
+		// if ( false === ( $addon_sections ) ) {
```
You probably want to re-enable this. :)

claudiulodro replied:
After that's ^ changed I give this the green light.

kovshenin commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -40,36 +40,97 @@ public static function get_featured() {
+		if ( $country ) {
+			$transient_name = $transient_name . '_country_' . $country;
+		}
+		return $transient_name;
```
I think we should hash these for two reasons.

First – there's a 191 character limit on `option_name` in the `wp_options` table, so searching for really long strings (or copy/paste errors, or search by description, mostly edge cases I guess...) will truncate names silently on most MySQL/Maria servers.

Second – with `$term` we're piping user input directly into `set_transient()`, probably not a huge deal since options in WordPress can actually be named `'; DROP TABLE wp_users; --` and still work fine, but hey. Better safe than SQLi.

What I would do is probably something along the lines of:

```
public static function build_transient_name( $args ) {
    return 'wc_addons_extensions_' . md5( json_encode( $args ) ) );
}
```

Then you can call it with `array( 'term' => 'foo', 'category' => 'shipping-methods' )` or whatever.

kovshenin commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -40,36 +40,97 @@ public static function get_featured() {
+			if ( ! is_wp_error( $raw_extensions ) ) {
+				$addons = json_decode( wp_remote_retrieve_body( $raw_extensions ) )->products;
+				if ( $addons ) {
+					set_transient( $transient_name, $addons, WEEK_IN_SECONDS );
```
We should be caching failures here too, for when WooCommerce.com is down.

Also, I'm not completely sold on caching results for actual search terms on the client side. I see the benefit of caching some things on the WooCommerce.com API side, when multiple different users with different sites search for the same thing in the same country and category.

But for *this* specific client-side cache to be effective, the user will have to go to the Extensions screen and perform a search. Then within a week, go to the Extensions screen and perform the exact same search. Sounds a bit too coincidental to me.

Caching categories is ok I feel, although if you think about it, what if a hot new featured extension is launched on WooCommerce.com, it has all the greatest features and all these other amazing things, and as a promo it costs only 7 cents, and we're running the promo for a week only. Some unlucky users will miss out on this great opportunity because of stale cache :)

kovshenin commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -40,36 +40,97 @@ public static function get_featured() {
+		// }
 
-		return apply_filters( 'woocommerce_addons_sections', $addon_sections );
+		apply_filters( 'woocommerce_addons_sections', $addon_sections );
```
Did you mean:

```
$addon_sections = apply_filters( ... );
```

kovshenin commented on includes/admin/views/html-admin-page-addons.php:
```diff
@@ -25,19 +25,41 @@
+
+		<?php if ( isset( $_GET['search'] ) ) : ?>
+			<h1 class="search-form-title" >
+				Showing Search results for:
```
Need to `__( 'Internationalize', 'this-string' );`

kovshenin commented on includes/admin/views/html-admin-page-addons.php:
```diff
@@ -25,19 +25,41 @@
+			<button type="submit">
+				<span class="dashicons dashicons-search"></span>
+			</button>
+			<input type="text" name="search" value="" placeholder="Enter a search term and press enter">
```
Need to internationalize this placeholder. Perhaps also set the value to the current search term if not empty?

belcherj committed:
Fixes from code review

belcherj committed:
Remove unused function

claudiulodro commented on includes/admin/views/html-admin-page-addons.php:
```diff
@@ -25,19 +25,45 @@
+			<input
+				type="text"
+				name="search"
+				value="<?php echo esc_html( $_GET['search'] ); ?>"
```
`<br /><b>Notice</b>:  Undefined index: search in <b>/srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce/includes/admin/views/html-admin-page-addons.php</b> on line <b>53</b><br />` if you just navigate to any section without searching or when you first navigate to the extensions screen.

claudiulodro replied:
Also I believe technically you want [esc_attr](https://developer.wordpress.org/reference/functions/esc_attr/) for attributes like `value`.

claudiulodro commented:
Not technically something you're working on in this PR, but if you're changing things around in this file will you please switch the `esc_attr` to `esc_url` on [this line](https://github.com/woocommerce/woocommerce/pull/17926/files#diff-8d82bee6e0d37845ad4c8bca62c9ecbbR96). [Ref this comment](https://developer.wordpress.org/reference/functions/esc_attr/#comment-2459)

belcherj committed:
Fix escaping

claudiulodro commented:
Excellent. Looks good to me! I'll merge this tomorrow unless anyone has pressing feedback.

kovshenin commented on includes/admin/views/html-admin-page-addons.php:
```diff
@@ -25,19 +25,45 @@
+		<?php if ( isset( $_GET['search'] ) ) : ?>
+			<h1 class="search-form-title" >
+				<?php _e( 'Showing Search results for:', 'woocommerce' ); ?>
+				<strong><?php echo esc_html( $_GET['search'] ); ?></strong>
```
The search term should really be a placeholder inside the translated string, which will allow RTL languages to move things around:

```
printf( __( 'Showing search results for: %s', 'woocommerce' ), '<strong>' . esc_html( $_GET['search'] ) . '</strong>' ) );
```

mikejolley commented:
@belcherj I ran out of time to sort it today, but in addition to the placeholder comment, take a look at the bar on the main featured page - to me it looks weird how some things have max width set, and others do not.

The bar, featured content, and grid of products should have equal width. Should be do-able with CSS?

belcherj commented:
@mikejolley I agree that some not having max width is weird but I think it is outside the scope of this commit.  Adding a max width to the featured page would mean that I need to add max width to the categories page which I think would be a detriment since it would show less extensions.  @garymurray what do you think the best path forward is?

belcherj committed:
Merge commit

belcherj committed:
Fix placeholder comment

belcherj committed:
Merge Commit

belcherj commented:
<img width="1335" alt="screen shot 2017-12-07 at 1 09 32 pm" src="https://user-images.githubusercontent.com/6817400/33731086-fffd86bc-db4f-11e7-9757-b025ec5e9d25.png">
<img width="1330" alt="screen shot 2017-12-07 at 1 09 48 pm" src="https://user-images.githubusercontent.com/6817400/33731087-000f31d2-db50-11e7-8025-9a73d953ec09.png">

mikejolley commented:
@belcherj Before I signed off I had a go making everything 100% width. It looked fine - I made the extensions 25% and min-width 300px. There were only a few glitches. We can log a separate issue if needed - it just needs cleaning up as it looks odd :p

claudiulodro commented:
If the functionality is working correctly let's merge this and create an issue for style tweaks. Those can be done after the beta if needed.

kloon commented:
Looks good, the failing tests is just PHPCS errors which we can fix at a later stage. @belcherj also worth noting that when searching for Jetpack it returns an empty block http://cld.wthms.co/51G1OT This is most likely something we would want to fix on the API side to always return a description? Or we can hide this within core when no description is set?

Styling issue can be done in the separate PR.

garymurray commented:
There should be one width that we stick to. We can address that as a separate fix though I think.
